client/llb/exec.go

```diff
@@ -142,6 +143,11 @@ func (e *ExecOp) Marshal(c *Constraints) (digest.Digest, []byte, *pb.OpMetadata,
 			e.meta.Env = e.meta.Env.AddOrReplace("SSH_AUTH_SOCK", e.ssh[0].Target)
 		}
 	}
+	if err := c.Caps.Supports(pb.CapExecMetaSetsDefaultPath); err != nil {
```

Comment:
What behavior makes more sense here if `c.Caps` is nil (eg. offline on a direct client with no filtering set). I think we could assume up-to-date daemon then.

Reply:
#641 requires the user to have known (and remembered) to pass `llb.WithCaps(caps)` to `Marshal` here. If they do not (i.e. all preexisting code) then `c.Caps` will be `nil` and they would be expecting the legacy behaviour -- i.e. CapExecMetaSetsDefaultPath not present so they get a default PATH set at this level.

There's a good argument that if `WithCaps` isn't used then `Marshall` should inject a set of caps which corresponds to the state at the point `WithCaps` was added, but that also wouldn't include `CapExecMetaSetsDefaultPath` so while it is a good idea IMHO it's a bit moot WRT this PR.

Reply:
> then c.Caps will be nil and they would be expecting the legacy behaviour -- i.e. CapExecMetaSetsDefaultPath not present so they get a default PATH set at this level.

I think they shouldn't get the default path. If they explicitly say that they have a set of caps that doesn't contain this we can skip it. This is more in line with the other features that just add the caps they use for validation instead of expecting that they are not supported if caps is not set. So I think this can just be wrapped in `if c.Caps != nil`.

Reply:
OK. I also interpreted this as implying I should drop `Allow `apicaps.CapSet.Supports` to be called on a nil CapSet.` from this PR, so I will do that too.
```diff
commit 0908e733f98eac440ac8f472483313f275ef016a
Author: Ian Campbell <ijc@docker.com>
Date:   Tue Sep 25 10:46:09 2018 +0100

    Allow `apicaps.CapSet.Supports` to be called on a nil CapSet.
    
    Signed-off-by: Ian Campbell <ijc@docker.com>

diff --git a/util/apicaps/caps.go b/util/apicaps/caps.go
index 9a661b50..9ead25ca 100644
--- a/util/apicaps/caps.go
+++ b/util/apicaps/caps.go
@@ -106,6 +106,9 @@ type CapSet struct {
 // Supports returns an error if capability is not supported
 func (s *CapSet) Supports(id CapID) error {
        err := &CapError{ID: id}
+       if s == nil || s.list == nil {
+               return err
+       }
        c, ok := s.list.m[id]
        if !ok {
                return errors.WithStack(err)
```